Feature/architecture

leoTiez opened the pull request:
The architecture is implemented according to the paper. However, I dropped the keras implementation and used 'pure' tensorflow instead, since it is more comprehensible what actually happens. Training works and the network can achieve around 85% accuracy. Note amazing, but the hyperparameters are not yet set properly. 

Moreover, I want to investigate whether it is necessary to implement the backpropagation through time algorithm, or whether the 'standard' backprop implementation is sufficient. However, I think this is a new feature and it might be more appropriate to put it in a separate branch

leoTiez committed:
Update: Changed architecture accrodingly to the paper

Moreover, changed loss metric to categorical crossentropy

leoTiez committed:
Update: Added transformation of labels to hot-1

Now, the preprocessing functions require the labels as additional
parameter. They will be converted to a hot-1 notation.

leoTiez committed:
Update: Tried to fix learning issue

Architecture was updated and changed according to the proposal
in the paper. However, the accuracy after learning is
as high as guessing by chance. It could not be figured out
why this is the case. Unfortunately, since keras is a high
level api it is hard to figure out what happens when.

leoTiez committed:
Update: Adde --user flag to the install requirements command

leoTiez committed:
Update: Implemented architecture in 'pure' tensorflow style.

Now the layers are defined using the standard tensorflow syntax
instead of using a high level api, such as Keras.
Results are promising so far. The architecture, however, has not
been properly tested yet.

leoTiez committed:
Update: Used division by 255 instead of normalization using std

Since some of the values were off while applying normalization
using std, the values are divided by 255 now

leoTiez committed:
Update: Implemented architecture proposed by the paper

The rcl, softmax layer as well as the pooling and the standard
convolutional layers are implemented and are working. Results
are promising so far. However, the accuracy indicates
that the global max pooling is not workin properly. It is still
necessary to investigate this issue.

leoTiez committed:
Update: Changed some parameters for more effective learning

Checking against the proposal in the paper led to the conclusion
that the global max layer is correctly implemented. However,
it is apparently not alway used.

Still missing: dropout layer

Open question: Is bptt still necessary?

leoTiez committed:
Update: Added dropout layers

leoTiez committed:
Update: Resolved todo in the comments

adrianc-a commented:
I'd prefer backprop over time on it's own branch

adrianc-a commented on CRNN.py:
```diff
@@ -193,35 +194,87 @@ def compute_output_shape(self, input_shape):
     model.compile(
         optimizer=tf.train.AdagradOptimizer(learning_rate=.1),
-        loss='sparse_categorical_crossentropy',
+        loss='kullback_leibler_divergence',
```
why have we changed this? the orig. paper uses cross-entropy.

leoTiez replied:
Was a test. I will change it. However, we should use the `RCNN_tf` implementation anyways.

adrianc-a commented on RCNN_tf.py:
```diff
@@ -0,0 +1,338 @@
+    return output
+
+
+def pooling_layer(
```
I get that you want to use 'pure' tf, but i feel like this introduces more potential for error. what is wrong with using [tf.nn.pool](https://www.tensorflow.org/api_docs/python/tf/nn/pool) ?

leoTiez replied:
Take a look some lines below :). This is just to provide a simple interface

adrianc-a commented on RCNN_tf.py:
```diff
@@ -0,0 +1,338 @@
+    return tf.reduce_max(input_data, axis=[1, 2])
+
+
+def softmax_layer(
```
again, what is the advantage/reason over `tf.nn.softmax`?

leoTiez replied:
Again, `tf.nn.softmax` is used. But we need to compute the linear transformation first.

adrianc-a commented on RCNN_tf.py:
```diff
@@ -0,0 +1,338 @@
+    return tf.nn.softmax(linear_trans), linear_trans
+
+
+def accuracy(
```
`tf.metrics.accuracy` ?

leoTiez replied:
I will take a look on that.

adrianc-a commented on RCNN_tf.py:
```diff
@@ -0,0 +1,338 @@
+
+if __name__ == '__main__':
+    # Setting the parameters
+    input_shape_ = [None, 28, 28, 1]
```
realistically, all the problem specific dimensions need to be abstracted for now it's probably ok, but i guess keep it in mind.

leoTiez replied:
I don't understand what you mean. I what way abstracting it?

adrianc-a replied:
we should have separate main functions, and the network would have to have different input dimension, but i think we can save this for later.

leoTiez replied:
:+1:

adrianc-a commented on RCNN_tf.py:
```diff
@@ -0,0 +1,338 @@
+    input_shape_ = [None, 28, 28, 1]
+    output_shape_ = [None, 10]
+    learning_rate_ = .1
+    epochs_ = 5
```
also more a comment for the future, we should just save checkpoints and make this not so fixed.

adrianc-a commented on RCNN_tf.py:
```diff
@@ -0,0 +1,338 @@
+        global_max_.shape[-1].value,
+        output_shape_[1]
+    )
+
```
afaik, the architecture looks as it does in the paper, good.

adrianc-a commented on RCNN_tf.py:
```diff
@@ -0,0 +1,338 @@
+    flatten_ = tf.reshape(global_max_, [-1, global_max_.shape[-1].value])
+
+    # Softmax layer
+    result_, linear_trans_ = softmax_layer(
```
clever

adrianc-a commented on RCNN_tf.py:
```diff
@@ -0,0 +1,338 @@
+                "test accuracy: {:.3f}".format(test_acc_)
+
+        print "\nTraining complete!"
+        print sess_.run(accuracy_,
```
this can really be simplified using `tf.data.Dataset` which can handle shuffling, batching, repeating etc.

leoTiez replied:
I will take a look on that

adrianc-a commented on load_data.py:
```diff
@@ -34,19 +36,19 @@ def load_cifar(file_name, dtype, file_path='./data/cifar-10-batches-py/'):
-    R = cifar_data[:, 0:1024]
-    G = cifar_data[:, 1024:2048]
-    B = cifar_data[:, 2048:]
+    R = cifar_data[:, 0:1024] / 255.
```
not super pertinent to this pr, but i feel like we should standardize image preprocessing

leoTiez replied:
Depends on what you mean with standardizing. All Images of all data sets should be preprocessed the same way, I totally agree. We only need to do this consistently. Which normalization we use is less important I find.

adrianc-a replied:
yes, just that it should be the same across the datasets.

adrianc-a commented:
Also, this whole CRNN/RCNN naming scheme with two different main's is i think convoluted. since this is it's own branch and you're trying to fix CRNN.py, why don't we just overwrite that?

adrianc-a committed:
Merge branch 'develop' into feature/architecture

leoTiez commented:
I can remove the old implementation. However, I thought it would be nice to provide a Keras implementation to the community.

leoTiez committed:
Update: Changed accuracy metric back to cross entropy

leoTiez committed:
Update: Implement tensorflow dataset objects

Instead of using the feed_dict in the learning phase for passing
the training data, the tf dataset objects are used. This makes
learning more efficient and batch gradient descent easier.

Moreover, the provided tensorflow accurcy metric is used
instead of the own implemenetation.

adrianc-a commented on RCNN_tf.py:
```diff
@@ -169,28 +164,44 @@ def accuracy(
+    training_data_set_ = tf.data.Dataset.from_tensor_slices((
+        input_placeholder_,
+        output_placeholder_
+    )).repeat().batch(batch_size=batch_size_)
```
i think we should also shuffle it.

adrianc-a reviewed (approved):
let's tidy it up later.

leoTiez committed:
Update: Added shuffling of data
